components/text_recognition/browser/text_recognition_helper_win.cc

```diff
@@ -0,0 +1,100 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/text_recognition/browser/text_recognition_helper_win.h"
+
+#include <utility>
+
+#include "base/logging.h"
+#include "base/task/task_traits.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/text_recognition/browser/text_recognition.h"
+#include "content/public/browser/browser_thread.h"
+
+namespace text_recognition {
+
+TextRecognitionHelperWin::TextRecognitionHelperWin() = default;
+TextRecognitionHelperWin::~TextRecognitionHelperWin() = default;
+
+void TextRecognitionHelperWin::GetTextFromImage(
+    const SkBitmap& image,
+    base::OnceCallback<void(const std::vector<std::string>&)> callback) {
+  callback_ = std::move(callback);
+  image_ = image;
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::MayBlock(), base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN},
+      base::BindOnce(&GetAvailableRecognizerLanguages),
+      base::BindOnce(&TextRecognitionHelperWin::OnGetAvailableLanguages,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void TextRecognitionHelperWin::SetResultText(
+    const std::vector<std::string>& text) {
+  if (text.empty()) {
+    return;
+  }
+
+  if (best_result_.empty()) {
+    best_result_ = text;
+  }
+
+  // Determine which one is more better result between |text| and
+  // |best_result_|. We choose one that has more longer detected string.
+  int text_string_size = 0;
+  for (const auto& s : text) {
+    text_string_size += s.size();
+  }
+
+  int best_result_string_size = 0;
+  for (const auto& s : best_result_) {
+    best_result_string_size += s.size();
+  }
+
+  if (text_string_size > best_result_string_size) {
+    best_result_ = text;
+  }
+}
+
+void TextRecognitionHelperWin::OnGetTextFromImage(
+    const std::vector<std::string>& text) {
+  CHECK(content::BrowserThread::CurrentlyOn(content::BrowserThread::UI));
+
+  remained_language_detection_request_count_--;
+
+  SetResultText(text);
+
+  if (remained_language_detection_request_count_ == 0) {
```

Comment:
what happens if some lang request will be crashes or lost?

Reply:
It's good question. I think `TextDetectionImplWin` could have timer for that.
If timer expires, `TextDetectionImplWin` could give empty result instead of waiting.

Reply:
If `OcrEngine.RecognizeAsync()` doesn't return in an async way, our dialog will be waiting that result forever.
I'm thinking to add timer for it but not sure how much we should wait because it could be long based on the amount of text in the image. Also, user could destroy the dialog at any time by using `esc`. and could request again.
Also upstream also doesn't consider this situation. How about going as-is?